### PR TITLE
Reactor unobserve

### PR DESCRIPTION
**Pull request by colindresj**
Creates reactor#unobserve and KeyPath.eql, which tests if two keypaths are the same.

### CODE REVIEW

**Comment by jordangarcia** [comment]
I will check this out tomorrow and play around.  Given tests are passing it LGTM but I'd like to mess around with it first hand before merging.

**Comment by colindresj** [comment]
Cool, not sure on the `KeyPath.eql` name tbh. Maybe `Keypath.same`, or something like that.

File: src/key-path.js [hunk 1]
```diff
@@ -6,9 +6,36 @@ var isFunction = require('./utils').isFunction
...
+
+  var len
+  var i = len = keyPath.length
+
+  if (len !== otherKeyPath.length) {
+    return false
+  }
+
+  while (i--) {
+    if (keyPath[i] !== otherKeyPath[i]) {
```
Comment (jordangarcia) [inline src/key-path.js]: This needs to be recursive since getters can be composed of other getters.

It may be good to rely on the getetr hashCode, although I am a bit hesitant as others have pointed out that there is a possibility of collisions.

Comment (colindresj) [inline src/key-path.js]: I might be missing something, but this is just for comparing array keypaths of `Array [ String, ...]`. It uses `isKeyPath` to make sure of that.

The actual getter comparison is an `===` reference check, since we have access to the observer entry: https://github.com/optimizely/nuclear-js/pull/115/files#diff-889e806243cf10b2e10689e9632f17a2R132.

[Here](https://github.com/optimizely/nuclear-js/pull/115/files#diff-7e02ba8688bee6af79401977651e94b9R284)'s the test for getters, using `totalGetter`, which is composed of other getters.

File: src/reactor.js [hunk 2]
```diff
@@ -92,6 +92,48 @@ class Reactor {
...
+        }
+
+        return null
+      }
+
+      return entry.unwatchFn
+    }
+
+    each(this.__changeObserver.__observers.map(entry => {
+      if (keyPath && KeyPath.eql(keyPath, Getter.getDeps(entry.getter)[0])) {
```
Comment (jordangarcia) [inline src/reactor.js]: Kind of confused what this line is doing.

If the getter is a keyPath then we are comparing against the first dep of the `enter.getter`?

It seems like this should be something like:

```js
const isKeyPath = keyPath.isKeyPath()

// unwatchifShould invokes entry.unwatchFn if found

each(this.__changeObserver.__observers.map(entry => {
  if (isKeyPath && KeyPath.isKeyPath(entry.getter) && KeyPath.eql(getter, entry.getter)) {
    return unwatchIfShould(entry)
  } 
  if (entry.getter === getter) {
    return unwatchIfShould(entry)
  }
}), unwatchFn => unwatchFn && unwatchFn())
```

Also are you using `map` and `each` as to not mutate the `__observers` map while iterating?

Lastly I think this logic should be in the `ChangeObserver` under `__changeObserver.unobserve`.  IMO the reactor shouldn't have knowledge of the inner workings of `ChangeObserver`

Comment (colindresj) [inline src/reactor.js]: Good point. I'll update and move logic to the change observer.

**Comment by bhamodi** [comment]
Hey @colindresj, what's the status of this PR?

**Comment by colindresj** [comment]
@bhamodi sorry slipped my mind I had this PR open. I'll work on the stuff I said I would shortly, and will push. Thanks for the reminder

File: src/change-observer.js [hunk 3]
```diff
@@ -61,20 +64,69 @@ class ChangeObserver {
...
+   * Calls the unwatchFn for each change observer associated with the passed in
+   * keypath or getter. If a handler function is passed as well, then the
+   * unwatchFn is called only for the change observer that holds that handler.
+   *
+   * UnwatchFns are called up the dependency chain.
+   *
+   * @param {KeyPath|Getter} getter
+   * @param {function} handler
+   */
+  unwatch(getter, handler) {
```
Comment (colindresj) [inline src/change-observer.js]: Moved all the logic into the changeObserver

File: src/change-observer.js [hunk 4]
```diff
@@ -61,20 +64,69 @@ class ChangeObserver {
...
+
+      return null
     }
+
+    var isKeyPath = KeyPath.isKeyPath(getter)
+
+    // Collects all the unwatchFns that need to be called, without invoking
+    // them, so as to not mutate the observers collection, then invokes each.
+    each(this.__observers.map(entry => {
+      if (isKeyPath && Getter.wasKeyPath(entry.getter)) {
```
Comment (colindresj) [inline src/change-observer.js]: Checking on whether both the passed getter is a keypath and if the entry's registered getter was originally a keypath, in which case we can access the original keypath using `entry.getter[0]`, since keypaths are transformed to getters using `[keypath, identity]`

File: src/change-observer.js [hunk 5]
```diff
@@ -61,20 +64,69 @@ class ChangeObserver {
...
+        if (KeyPath.same(getter, entry.getter[0]) || KeyPath.isUpstream(getter, entry.getter[0])) {
+          return unwatchIfShould(entry)
+        }
+      }
+
+      if (entry.getter === getter) {
```
Comment (colindresj) [inline src/change-observer.js]: If passed getter is a getter, we can do a simple referential equality check

File: src/change-observer.js [hunk 6]
```diff
@@ -61,20 +64,69 @@ class ChangeObserver {
...
+      return null
     }
+
+    var isKeyPath = KeyPath.isKeyPath(getter)
+
+    // Collects all the unwatchFns that need to be called, without invoking
+    // them, so as to not mutate the observers collection, then invokes each.
+    each(this.__observers.map(entry => {
+      if (isKeyPath && Getter.wasKeyPath(entry.getter)) {
+        if (KeyPath.same(getter, entry.getter[0]) || KeyPath.isUpstream(getter, entry.getter[0])) {
```
Comment (colindresj) [inline src/change-observer.js]: Will unwatch when the passed keypath is the same as the entry's, or if it is upstream the entry's. IE: `['key']` is upstream `['key', 'path']`.

**Comment by colindresj** [comment]
@bhamodi @jordangarcia Long time coming, but updated this PR. Added even more tests to cover some potential edge cases, and updated logic based on suggestions as much I could.

**Comment by bhamodi** [comment]
@colindresj - Sorry for the delay in getting back to you. This PR now requires a rebase.
@jordangarcia - Could you review this PR one more time before we merge it in?

**Comment by colindresj** [comment]
Rebased @bhamodi

**Comment by bhamodi** [comment]
Thanks for doing this @colindresj! This LGTM.
Waiting to hear from @jordangarcia and then we'll merge it in.

**Comment by bhamodi** [comment]
This has been implemented in v1.2.0 by @jordangarcia. Thanks @colindresj.
